src/core/server/application/application_context_provider.ts

```diff
@@ -0,0 +1,181 @@
+
+  private async getUiPlugins(): Promise<ConfiguredPlugin[]> {
+    return Promise.all(
+      [...this.plugins.uiPlugins.public].map(async ([id, plugin]) => {
+        const config = await this.plugins.uiPluginConfigs
+          .get(id)
+          ?.pipe(take(1))
+          .toPromise();
+
+        return { id, plugin, config: config || {} };
+      })
+    );
+  }
+
+  private async getMetadata(
+    injectedOverrides: PluginVariables,
+    includeUserProvidedConfig: boolean
+  ) {
+    const [uiPlugins, uiSettings, userSettings] = await Promise.all<
+      ConfiguredPlugin[],
+      PluginVariables,
+      UserVariables
+    >([
+      this.getUiPlugins(),
+      this.uiSettings.getAll(),
+      includeUserProvidedConfig ? this.uiSettings.getUserProvided() : Promise.resolve({}),
+    ]);
+
+    return {
+      uiPlugins,
+      uiSettings,
+      userSettings,
+      vars: mergeVariables(
+        uiPlugins.map(({ config }) => config),
+        injectedOverrides
+      ),
+    };
+  }
+
+  private async getVariables(
+    pluginId: string,
+    injectedOverrides: PluginVariables,
+    includeUserProvidedConfig: boolean
+  ) {
+    const { uiPlugins, uiSettings, userSettings, vars } = await this.getMetadata(
+      injectedOverrides,
+      includeUserProvidedConfig
+    );
+    const basePath = this.http.basePath.get(this.request);
+    // TODO: Use full implementation once https://github.com/elastic/kibana/pull/51438 lands:
+    const capabilities = defaultCapabilities;
+
+    return {
+      uiPublicUrl: `${basePath}/ui`,
+      bootstrapScriptUrl: `${basePath}/bundles/app/${pluginId}/bootstrap.js`,
+      i18n: i18n.translate,
+      locale: i18n.getLocale(),
+      darkMode: userSettings['theme:darkMode']?.userValue ?? false,
+      injectedMetadata: {
+        version: this.http.server.version,
+        buildNumber: this.env.packageInfo.buildNum,
+        branch: this.env.packageInfo.branch,
+        basePath,
+        env: this.env,
+        legacyMode: false,
+        i18n: {
+          translationsUrl: `${basePath}/translations/${i18n.getLocale()}.json`,
+        },
+        csp: this.http.csp,
+        vars,
+        uiPlugins,
+        capabilities,
+        legacyMetadata: {
+          app: {},
+          bundleId: `app:${pluginId}`,
+          nav: capabilities.navLinks,
+          env: this.env,
+          version: this.env.packageInfo.version,
+          branch: this.env.packageInfo.branch,
+          buildNum: this.env.packageInfo.buildNum,
+          buildSha: this.env.packageInfo.buildSha,
+          serverName: this.http.server.name,
+          devMode: this.env.mode.dev,
+          basePath,
+          uiSettings: {
+            defaults: uiSettings,
+            user: userSettings,
+          },
+        },
+      },
+    };
+  }
+
+  public async render(id = 'core', injectedVars = {}, includeUserProvidedConfig = true) {
+    const headers = {
+      'content-security-policy': createCspRuleString(this.http.csp.rules),
```

Comment:
Looks like it's rendering service responsibility to test that CSP can be configured via config file